Adds colored logging output using logger.copyline

jhaux opened the pull request:
This functionality is supposed to be used for outputs containing code
snippets that can be copied like edeval commands. The color makes it so
that these lines do not get lost and improve the workflow speed.

jhaux committed:
:tada: Adds colored logging output using logger.copyline

This functionality is supposed to be used for outputs containing code
snippets that can be copied like edeval commands. The color makes it so
that these lines do not get lost and improve the workflow speed.

pesser commented on edflow/custom_logging.py:
```diff
@@ -423,21 +423,63 @@ def _create_logger(name, out_dir, level=logging.INFO):
         logger = logging.getLogger(name)
         logger.setLevel(level)
 
+        logger.copyline = lambda msg: logger.log(22, msg)
```
do we want to use something specific like `copyline` or something more generic like `colored_info`?
Why is `levelno` 22 used? anything special about that?

jhaux replied:
> do we want to use something specific like copyline or something more generic like colored_info?

Agreed!

> Why is levelno 22 used? anything special about that?

During formatting we can access the level and use this info to switch to color mode.

pesser commented on edflow/custom_logging.py:
```diff
@@ -423,21 +423,63 @@ def _create_logger(name, out_dir, level=logging.INFO):
             fh = logging.FileHandler(filename=os.path.join(out_dir, "log.txt"))
 
             fmt_string = "[%(levelname)s] [%(name)s]: %(message)s"
             formatter = logging.Formatter(fmt_string)
```
is the `formatter` still needed or can we delete it?

jhaux replied:
Yep, should be deleted.

pesser commented on edflow/custom_logging.py:
```diff
@@ -423,21 +423,63 @@ def _create_logger(name, out_dir, level=logging.INFO):
+            "msg": record.msg,
+        }
+
+        if record.levelno == 22:
```
can we make this formatter generic enough so that we can also do something like `colored_info("red", msg)` for custom colored logging output?

jhaux replied:
I like the idea! I will have a look at it.

jhaux replied:
Added in my latest [commit](https://github.com/pesser/edflow/pull/230/commits/168b3725c04104ba16033d84edb47f18705c9add)

jhaux committed:
:fire: Removes unused lines and changes copyline to colored_info

jhaux committed:
:tada: Adds more general color logging

Now the edflow loggers (get_logger(name)) have the method `color_info`
which logs at info level a line in the color as supplied by the keyword
arg `color=<color>`.

jhaux commented:
This is what it now looks like:
![color_print](https://user-images.githubusercontent.com/9572598/75441335-c9c23500-595d-11ea-9f95-42db0815c536.png)

jhaux committed:
:tada: Even more general: no extra log command, only color kwarg

As it turns out, logging has already a lot of cool functionality, which
can be easily leveraged to get the behavior we want. Now any log level
command can be used with a color keyword argument to make it fancy.

jhaux committed:
:white_check_mark: Colored logging test

Simply running a lot of possible logging commands.

jhaux commented:
Things are even more general now. I redid the colored_info wrapper and realized that it is far easier to wrap the underlying` _log` method. Now we can color any log command!

![color_print](https://user-images.githubusercontent.com/9572598/75457898-59c3a700-597d-11ea-8d04-7c415130cc56.png)

jhaux committed:
:bug: Fixes wrong handling of no color passed and adds test

pesser reviewed (approved):
love it!
